packages/client/wallets/smart-wallet/src/blockchain/wallets/SendTransactionService.test.ts

```diff
@@ -0,0 +1,140 @@
+import { SmartAccountClient } from "permissionless";
+import { SmartAccount } from "permissionless/accounts";
+import { EntryPoint } from "permissionless/types";
+import {
+    BaseError,
+    Chain,
+    ContractFunctionRevertedError,
+    PublicClient,
+    SimulateContractReturnType,
+    TransactionReceipt,
+    Transport,
+    zeroAddress,
+} from "viem";
+import { beforeEach, describe, expect, it, vi } from "vitest";
+import { mock } from "vitest-mock-extended";
+
+import {
+    EVMSendTransactionError,
+    EVMSendTransactionExecutionRevertedError,
+    SendTransactionService,
+} from "./SendTransactionService";
+
+function makeMockError<E extends Error, F extends object>(error: E, fields?: F): E & F {
+    const mockError = Object.create(error);
+    Object.assign(mockError, fields);
+    return mockError;
+}
+
+describe("SendTransactionService", () => {
+    const mockPublicClient = mock<PublicClient>();
+    const mockAccountClient = mock<SmartAccountClient<EntryPoint, Transport, Chain, SmartAccount<EntryPoint>>>();
+    const sendTransactionService = new SendTransactionService(mockPublicClient);
+
+    beforeEach(() => {
+        vi.resetAllMocks();
+    });
+
+    it("Throws EVMSendTransactionExecutionRevertedError when a transaction simulation fails", async () => {
+        const mockError = makeMockError(BaseError.prototype, { walk: vi.fn() });
+        const mockRevertError = new ContractFunctionRevertedError({
+            abi: [],
+            functionName: "mockFunction",
+            message: "mockMessage",
+        });
+        mockError.walk.mockReturnValue(mockRevertError);
+        mockPublicClient.simulateContract.mockRejectedValue(mockError);
+        await expect(
+            sendTransactionService.sendTransaction(
+                {
+                    address: zeroAddress,
+                    abi: [],
+                    functionName: "mockFunction",
+                    args: [],
+                },
+                mockAccountClient
+            )
+        ).rejects.toThrow(EVMSendTransactionExecutionRevertedError);
+    });
+
+    it("Throws EVMSendTransactionExecutionRevertedError when a transaction reverts on chain", async () => {
+        const mockReceipt = mock<TransactionReceipt>();
+        mockReceipt.status = "reverted";
+        mockPublicClient.waitForTransactionReceipt.mockResolvedValueOnce(mockReceipt);
+        await expect(
+            sendTransactionService.sendTransaction(
+                {
+                    address: zeroAddress,
+                    abi: [],
+                    functionName: "mockFunction",
+                    args: [],
+                },
+                mockAccountClient
+            )
+        ).rejects.toThrow(EVMSendTransactionExecutionRevertedError);
+    });
```

Comment:
You could make these more robust by also checking the stage like you do in the test below